TundraTD/Assets/Scripts/Spells/Grimoire.cs

```diff
@@ -29,39 +31,42 @@ static Grimoire()
         /// <summary>
         /// Turns elements into a spell if it's available.
         /// </summary>
-        /// <param name="elements">Elements list.</param>
+        /// <param name="hitInfo"></param>
+        /// <param name="testMostElement">Manually set mostelement for debug</param>
         /// <returns>Created spell which is ready to cast.</returns>
-        public static MagicSpell TurnElementsIntoSpell(List<BasicElement> elements, Vector3 castPosition)
+        public static void TurnElementsIntoSpell(RaycastHit hitInfo, BasicElement testMostElement=BasicElement.None)
         {
-            BasicElement? mostElement = elements.GroupBy(x => x).FirstOrDefault(x => x.Count() >= 3)?.Key;
-
-            if (!mostElement.HasValue)
-                return null;
+            // Try to apply debug value
+            var mostElement = testMostElement;
+            if (mostElement == BasicElement.None) // And if it is none, then try to get it from the deck
+                mostElement = PlayerDeck.CurrentMostElement;    
+            // And don't move forward if there is no element atm
+            if (mostElement == BasicElement.None) return;
 
+            var elements = PlayerDeck.DeckElements.ToList();
             elements.Sort();
-            int startMostIndex = elements.IndexOf(mostElement.Value); 
+            
+            int startMostIndex = elements.IndexOf(mostElement); 
             var remainingElements = elements.Where((x, i) => x != mostElement || i >= startMostIndex + 3);
 
-            if (!SpellTypes.TryGetValue(mostElement.Value, out Type spellType))
-                return null;
-            
+            if (!SpellTypes.TryGetValue(mostElement, out Type spellType)) return;
+
             var spellObject = SpellInitializers[(int)Math.Log((int)mostElement, 2)];
             
             // don't call any spell if this spell was not implemented (for alpha purposes only)
             // TODO: Remove
-            if (spellObject is null)
-                return null;
-            
+            if (spellObject is null) return;
+
             var spell = Object.Instantiate(spellObject);
             
             foreach (var prop in spellType.GetProperties(BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Public))
             foreach (var attr in prop.GetCustomAttributes<UpgradeablePropertyAttribute>(true))
             foreach (var element in remainingElements)
                 attr.TryUpgradeProperty(element, prop, spell);
             
+            PlayerDeck.DeckElements.Clear();
             spell.SpellCameraLock += HandleSpellCameraLock;
-            spell.ExecuteSpell();
-            return spell;
+            spell.ExecuteSpell(hitInfo);
```

Comment:
If I remember, you are clearing player deck many times per one cast. It is cleared in the `SpellCaster` script and here. Is it necessary?

Reply:
Removed one extra line in 355a6bc